daedalus/src/redemption_key_validation.js

```diff
@@ -0,0 +1,208 @@
+"use strict";
+
+// This functions mirrors purescript variants from daedalus/src/Daedalus/ClientApi.purs
+// This is a part of rewrite from purescript into js and depricating purescript bits
+
+const bs58 = require("bs58");
+
+// Decode a Base64-encoded string using Node's `Buffer` API. Its trying to mimic atob functionality available in browsers.
+// This runs default node facility to decode base64 but additionally runs base64 validation code.
+// Note that atob under the hood runs validation. See example:
+//  > atob("====")
+//  InvalidCharacterError: String contains an invalid character
+//  > atob("aa==")
+//  "i"
+//  > atob("aa=")
+//  InvalidCharacterError: String contains an invalid character
+//
+//  Where node's implementation doesn't:
+//  > Buffer.from("====", "base64").toString("utf-8")
+//  ''
+//  > Buffer.from("aa==", "base64").toString("utf-8")
+//  'i'
+//  > Buffer.from("aa=", "base64").toString("utf-8")
+//  'i'
+//
+// And that's the reason this funciton exists: to extend Node's base64 decode with validity check.
+function decodeNode (str) {
+    var result;
+
+    // Check that the input string is a valid Base64-encoded string as Node.js
+    // decided that it would be a good idea to NOT throw on invalid input strings
+    // but return an empty buffer instead which cannot be distinguished from the
+    // empty string case.
+    var reEmptyString = "^$";
+    var leadingQuanta = "^([A-Za-z0-9+/]{4})*";
+    var finalQuantum =
+        "([A-Za-z0-9+/]{4}|[A-Za-z0-9+/]{3}(?:=)?|[A-Za-z0-9+/]{2}(?:=){0,2})$";
+    var reValidBase64 =
+        new RegExp([reEmptyString, "|", leadingQuanta, finalQuantum].join(""));
+
+    if (!reValidBase64.test(str)) { throw new Error("Invalid input string");}
+    result = Buffer.from(str, "base64").toString("utf-8");
+
+    return result;
```

Comment:
My only minor nitpick here is the use of regexes which looks like an hairy approach (although your excellent code coverage is making up for that). Let me see if I can propose something slightly different: if it's not feasible, we simply go back to your regex-based approach and I will be happy with that :)

OK, so here is my thought process: You say we cannot use the `Buffer` API because it is too lenient with malformed input (which is, it won't throw an error, but produce a "chopped" string), so here is an idea about an invariant: "An input is a valid base64-encoded string if and only if, after converting it to base64 and then to `ascii`, it preserves its original length".

To put it differently, we could write something similar (pardon my broken Node.js / JS skills):

```javascript
function decodeNode(original) {
  /* Note how we do not need to convert to UTF-8. RFC 4648 describes the Base64 Alphabet as:
  "A 65-character subset of US-ASCII is used, enabling 6 bits to be
   represented per printable character.  (The extra 65th character, "=",
   is used to signify a special processing function.)"
   */
  var converted    = Buffer.from(original, "base64").toString("ascii");

  // Note how `Buffer.from` performs a copy, so we can still access the old string size.
  vas isValidBase64 = converted.size() === original.size();
  if (!isValidBase64) throw new Error("Invalid input string");
  
  return converted;
}
```

Do you think I am being naive or this could actually work?

Reply:
> we simply go back to your regex-based approach

Note that its not *my* regex based approach - but its a copy paste from https://github.com/menelaos/purescript-b64/blob/master/src/Data/String/Base64.js#L11 . My modification was just to remove `Either` constructors.

I was trying to have exact implementation as before - not to introduce any regression.

My line of thought is - should we  modify implementation (as with your proposal we are doing that)?

> converting it to base64 and then to ascii, it preserves its original length

this sounds plausible, and I will try it out. Note what I have commented above though - this might and probably will change the implementation (we might just not care about that)

Reply:
Ah, interesting! (And to be 100% clear I'm not pointing fingers here, just trying to reason together if there is a better way to handle this 😉 ).

You make a valid point, this might change the implementation but hopefully not the end result (and I hope your tests will catch any regression). Let's see how it goes, and worst come to worst we will make a note in the code saying this is a direct porting on the PS code so the implementation is not any more fragile than the original 😉